pkg/driver/credential.go

```diff
@@ -0,0 +1,187 @@
+	k8sv1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/klog/v2"
+)
+
+type authenticationSource = string
+
+const (
+	// This is when users don't provide a `authenticationSource` option in their volume attributes.
+	// We're defaulting to `driver` in this case.
+	authenticationSourceUnspecified authenticationSource = ""
+	authenticationSourceDriver      authenticationSource = "driver"
+	authenticationSourcePod         authenticationSource = "pod"
+)
+
+const defaultHostPluginDir = "/var/lib/kubelet/plugins/s3.csi.aws.com/"
+
+const serviceAccountTokenAudienceSTS = "sts.amazonaws.com"
+
+const serviceAccountRoleAnnotation = "eks.amazonaws.com/role-arn"
+
+type credentialProvider struct {
+	client k8sv1.CoreV1Interface
+}
+
+func (c *credentialProvider) provide(ctx context.Context, req *csi.NodePublishVolumeRequest) (*MountCredentials, error) {
+	authenticationSource := req.VolumeContext[volumeCtxAuthenticationSource]
+	switch authenticationSource {
+	case authenticationSourcePod:
+		return c.provideFromPod(ctx, req)
+	case authenticationSourceUnspecified, authenticationSourceDriver:
+		return c.provideFromDriver()
+	default:
+		return nil, fmt.Errorf("unknown `authenticationSource`: %s, only `driver` (default option if not specified) and `pod` supported", authenticationSource)
+	}
+}
+
+func (c *credentialProvider) provideFromDriver() (*MountCredentials, error) {
+	klog.V(4).Infof("NodePublishVolume: Using driver identity")
+
+	hostPluginDir := hostPluginDirWithDefault()
+	hostTokenPath := path.Join(hostPluginDir, "token")
+
+	return &MountCredentials{
+		AccessKeyID:     os.Getenv(keyIdEnv),
+		SecretAccessKey: os.Getenv(accessKeyEnv),
+		SessionToken:    os.Getenv(sessionTokenEnv),
+		Region:          os.Getenv(regionEnv),
+		DefaultRegion:   os.Getenv(defaultRegionEnv),
+		WebTokenPath:    hostTokenPath,
+		StsEndpoints:    os.Getenv(stsEndpointsEnv),
+		AwsRoleArn:      os.Getenv(roleArnEnv),
+	}, nil
+}
+
+func (c *credentialProvider) provideFromPod(ctx context.Context, req *csi.NodePublishVolumeRequest) (*MountCredentials, error) {
+	klog.V(4).Infof("NodePublishVolume: Using pod identity")
+
+	tokensJson := req.VolumeContext[volumeCtxServiceAccountTokens]
+	if tokensJson == "" {
+		klog.Error("`authenticationSource` configured to `pod` but no service account tokens are received. Please make sure to enable `podInfoOnMountCompat`, see TODO")
+		return nil, status.Error(codes.InvalidArgument, "Missing service account tokens")
+	}
+
+	var tokens map[string]*Token
+	if err := json.Unmarshal([]byte(tokensJson), &tokens); err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "Failed to parse service account tokens: %v", err)
+	}
+
+	stsToken := tokens[serviceAccountTokenAudienceSTS]
+	if stsToken == nil {
+		klog.Errorf("`authenticationSource` configured to `pod` but no service account tokens for %s received. Please make sure to enable `podInfoOnMountCompat`, see TODO", serviceAccountTokenAudienceSTS)
```

Comment:
This could also happen if there's no service account linked or no annotation right? We should test this (not yet checked if we do) and have a better error message

Reply:
No, this would only happen if `tokenRequests` is missing in `CSIDriver` object. For all error messages in this file, I plan to create a document page for this feature and link all issues to there.